dot/state/block.go

```diff
@@ -542,10 +543,132 @@ func (bs *BlockState) GetSlotForBlock(hash common.Hash) (uint64, error) {
 	return types.GetSlotFromHeader(header)
 }
 
+func (bs *BlockState) loadHeaderFromDisk(hash common.Hash) (header *types.Header, err error) {
+	startHeaderData, err := bs.db.Get(headerKey(hash))
+	if err != nil {
+		return nil, fmt.Errorf("querying database: %w", err)
+	}
+
+	header = types.NewEmptyHeader()
+	err = scale.Unmarshal(startHeaderData, header)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshaling start header: %w", err)
+	}
+
+	if header.Empty() {
+		return nil, fmt.Errorf("%w: %s", chaindb.ErrKeyNotFound, hash)
+	}
+
+	return header, nil
+}
+
+// Range returns the sub-blockchain between the starting hash and the
+// ending hash using both block tree and disk
+func (bs *BlockState) Range(startHash, endHash common.Hash) (hashes []common.Hash, err error) {
+	if bs.bt == nil {
+		return nil, errNilBlockTree
+	}
+
+	endHeader, err := bs.loadHeaderFromDisk(endHash)
+	if errors.Is(err, chaindb.ErrKeyNotFound) {
+		// end hash is not in the disk so we should lookup
+		// block that could be in memory and in the disk as well
+		return bs.retrieveRange(startHash, endHash)
+	} else if err != nil {
+		return nil, fmt.Errorf("retrieving end hash from disk: %w", err)
+	}
+
+	// end hash was found in the disk, that means all the blocks
+	// between start and end can be found in the disk
+	return bs.retrieveRangeFromDisk(startHash, endHeader)
+}
+
+func (bs *BlockState) retrieveRange(startHash, endHash common.Hash) (hashes []common.Hash, err error) {
+	inMemoryHashes, err := bs.bt.SubBlockchain(startHash, endHash)
+	if err != nil {
+		return nil, fmt.Errorf("retrieving range from in-memory blocktree: %w", err)
+	}
+
+	firstItem := inMemoryHashes[0]
+
+	// if the first item is equal to the startHash that means we got the range
+	// from the in-memory blocktree
+	if firstItem == startHash {
+		return inMemoryHashes, nil
+	}
+
+	// since we got as many blocks as we could from
+	// the block tree but still missing blocks to
+	// fulfil the range we should lookup in the
+	// disk for the remaining ones, the first item in the hashes array
+	// must be the block tree root that is also placed in the disk
+	//  so we will start from its parent since it is already in the array
+	blockTreeRootHeader, err := bs.loadHeaderFromDisk(firstItem)
+	if err != nil {
+		return nil, fmt.Errorf("loading block tree root from disk: %w", err)
+	}
+
+	startingAtParentHeader, err := bs.loadHeaderFromDisk(blockTreeRootHeader.ParentHash)
+	if err != nil {
+		return nil, fmt.Errorf("range end should be in database: %w", err)
+	}
+
+	inDiskHashes, err := bs.retrieveRangeFromDisk(startHash, startingAtParentHeader)
```

Comment:
nit

```suggestion
	inDatabaseHashes, err := bs.retrieveRangeFromDisk(startHash, startingAtParentHeader)
```

also no idea why, but we usually say "on disk" instead of "in disk".... Although it's more "in the database" than "on the database" :laughing: